---
title: "De-duper Part 1"
author: "Nicki Zavoshy"
date: "10/15/2017"
output: html_document
---

#### Define the problem 
- The problem here is with PCR duplicates. This issue evolves when one or a set of transcipt preferentially gets duplicated during the PCR amplification step. This effectively eliminates the ability to quantify expression levels of the original RNA transcripts. By removing the PCR duplicates, the relative expression levels seen are easier to understand and the bias is decreased. This adds robustness to the data, and allows for more accurate downstream analysis. 

#### What do we need to do to discard PCR duplicates?
- These duplicates can be recognized by having the same information in a SAM file for multiple alignments. UMIs are used to differentiate sequences from each other. For our purposes, UMIs will be matched exactly and it is unimportant if there are uncalled bases or sequencing error. Only correct UMI containing reads will be retained. The chromsome position and starting position will also be identical between duplicates. Unfortunatly, soft clipping makes it a little more tricky than just finding repeat start positions and chromsome/scaffold number. With soft clipping, the basepair start position becomes 'artifically' shifted right since soft clipped portions do not align to the reference at all but the rest of the sequence does. To compensate for this, we need to check if softclipping has occurred and adjust the starting position accordingly. Then, we can see if there are PCR duplicates that may have just been soft clipped. Soft clipping only occurs at the beginning and end of a sequence. Therefore, even though there is soft clipping that could happen at the end, we are really only interested in soft clipping at the beginning since the start position is all that we care about right now. We are also interested in the strand orientation, denoted by the bitwise flag of either 0 or 16. We only want to retain one record per combination of:

    1. Strand orientation
    2. Known UMI
    3. Chromsome number
    4. left-most mapping starting position(basepair starting position)

#### Developing some Pseudocode


- begin by sorting the sam file with samtools sort. This will organize the reads by chromsome and starting position on the command line
- initiate an empty list
- loop through values of known UMIs and add those to the empty list above
- initiate an empty dictionary
- set a variable called previous chromsome to 1 so that we can compare chromosome values later
- loop through each line of the sam file
    - seperate the sam file into parts and save each of the following 4 components: UMI, strand value(0 or 16), chromsome and start position with soft 
    - look at the cigar string
    - if there has been soft clipping, then determine by how much and adjust the start position accordingly and overwrite this value
    - check if the UMI is in the list of known UMIs, if it is then continue on
          - check if the chromsome is equal to the previous value and if it is then continue on
          - set current chromosome value equal to past chromosome value
          - generate a tuple of this information
          - check if the tuple is in the dictionary, if not then add it to the dictionary with a count of 1 and write that line to an output file
          - if it isnt then move on to the next line and do nothing else to this line


**Notes about this assignmnet for later**

deduper
- umis will not be inline
- the umi will be in the header of that sequence 
- we need to check the header
- headr:line:more info: UMI
- If you find an UMI now that does not exactly match known UMIs then discard it (there are 96 known)
- plan for only single end 
* sort with samtools sort by chromosome and position 
* then look at the UMIs
*reading or writing files will take up the majority of your time 
** 

- check the UMI
- check the chromosome
- check the strand 
- corrected start position
- sort first 
